File: practice_questions1/cpp/finite_field/finite_field.cpp
#include "finite_field.h"

FieldElement::FieldElement(long long num, long long prime){
    if (num >= prime || num < 0){
        std::string error = "Num" + std::to_string(num) +
         "not in field range 0 to" + std::to_string(prime - 1);
        throw std::invalid_argument(error);
    }
    this->num = num;
    this->prime = prime;
}

std::string FieldElement::repr(){
    return "FieldElement_" + std::to_string(prime) + std::to_string(num);
}

bool FieldElement::operator!=(const FieldElement &other){
    return (this->num != other.num || this->num != other.prime);
}

bool FieldElement::operator==(const FieldElement &other){
    return this->num == other.num && this->prime == other.prime;
}

FieldElement *FieldElement::operator+(const FieldElement &other){
    if (this->prime != other.prime){
        throw std::invalid_argument("Cannot add two numbers in different Fields");
    }
    long long res = (this->num + other.num) % this->prime;
    return new FieldElement(res, this->prime);
}

FieldElement *FieldElement::operator-(const FieldElement &other){
    if (this->prime != other.prime){
        throw std::invalid_argument("Cannot add two numbers in different Fields");
    }
    long long res = (this->num - other.num) % this->prime;
    return new FieldElement(res, this->prime);
}

FieldElement *FieldElement::operator*(const FieldElement &other){
    if (this->prime != other.prime){
        throw std::invalid_argument("Cannot add two numbers in different Fields");
    }
    long long res = (this->num * other.num) % this->prime;
    return new FieldElement(res, this->prime);
}

FieldElement *FieldElement::Pow(const FieldElement &other,int number){
    long long res = 0;
    int i = 0;
    while (i < number){
        res = (res * other.num) % other.prime;
        i++;
    }
    return new FieldElement(res, other.prime);
}

File: practice_questions1/python/finite_field.py
class FieldElement:
    def __init__(self, num, prime):
        if num >= prime or num < 0:
            error = f"Num {num} not in field range 0 to {prime - 1}"
            raise ValueError(error)
        self.num = num
        self.prime = prime

    def __repr__(self):
        return f"FieldElement_{self.prime}({self.num})"

    def __ne__(self, other):
        return (self != other)

    
    def __eq__(self, other):
        if other is None:
            return False
        return self.num == other.num and self.prime == other.prime
    
    def __add__(self, other):
        if self.prime != other.prime:
            return TypeError('Cannot add two numbers in different Fields')
        num = (self.num + other.num) % self.prime
        return self.__class__(num, self.prime)
    
    def __sub__(self, other):
        if self.prime != other.prime:
            return TypeError('Cannot add two numbers in different Fields')
        num = (self.num - other.num) % self.prime
        return self.__class__(num, self.prime)


if __name__ == "__main__":
    a = FieldElement(7, 13)
    b = FieldElement(12, 13)
    c = FieldElement(6, 13)
    print(a+b==c)